planetiler-core/src/test/java/com/onthegomap/planetiler/pmtiles/PmtilesTest.java

```diff
@@ -133,7 +149,175 @@ void testRoundtripDirectorySlice() {
     in.add(new Pmtiles.Entry(1, 1, 1, 1));
     in.add(new Pmtiles.Entry(2, 3, 1, 1));
 
-    List<Pmtiles.Entry> out = Pmtiles.deserializeDirectory(Pmtiles.serializeDirectory(in, 1, 2));
+    List<Pmtiles.Entry> out = Pmtiles.directoryFromBytes(
+      Pmtiles.directoryToBytes(in, 1, 2));
     assertEquals(1, out.size());
   }
+
+  @Test
+  void testBuildDirectoriesRootOnly() throws IOException {
+    ArrayList<Pmtiles.Entry> in = new ArrayList<>();
+    for (int i = 0; i < 1000; i++) {
+      in.add(new Pmtiles.Entry(i, i * 100, 100, 1));
+    }
+    var result = WriteablePmtiles.makeDirectories(in);
+    assertEquals(1, result.numAttempts());
+  }
+
+  @Test
+  void testBuildDirectoriesLeavesNotTooSmall() throws IOException {
+    ArrayList<Pmtiles.Entry> in = new ArrayList<>();
+    for (int i = 0; i < 100000; i++) {
+      in.add(new Pmtiles.Entry(i, i * 100, 100, 1));
+    }
+    var result = WriteablePmtiles.makeDirectories(in);
+    assertTrue(result.leafSize() >= 4096, "entries in leaf: " + result.leafSize());
+  }
+
+  @Test
+  void testWritePmtilesSingleEntry() throws IOException {
+    var bytes = new SeekableInMemoryByteChannel(0);
+    var in = WriteablePmtiles.newWriteToMemory(bytes);
+
+    var config = PlanetilerConfig.defaults();
+    in.initialize(config, new TileArchiveMetadata(new Profile.NullProfile()), new LayerStats());
+    var writer = in.newTileWriter();
+    writer.write(new TileEncodingResult(TileCoord.ofXYZ(0, 0, 1), new byte[]{0xa, 0x2}, OptionalLong.empty()));
+
+    in.finish(config);
+    var reader = new ReadablePmtiles(bytes);
+    var header = reader.getHeader();
+    assertEquals(1, header.numAddressedTiles());
+    assertEquals(1, header.numTileContents());
+    assertEquals(1, header.numTileEntries());
+    assertArrayEquals(new byte[]{0xa, 0x2}, reader.getTile(0, 0, 1));
+    assertNull(reader.getTile(0, 0, 0));
+    assertNull(reader.getTile(0, 0, 2));
+
+    Set<TileCoord> coordset = reader.getAllTileCoords().stream().collect(Collectors.toSet());
+    assertEquals(1, coordset.size());
+  }
+
+  @Test
+  void testWritePmtilesToFileWithMetadata(@TempDir Path tempDir) throws IOException {
+    var in = WriteablePmtiles.newWriteToFile(tempDir.resolve("tmp.pmtiles"));
```

Comment:
Even though this is a test, might still be good to put this in a try-with-resources block to close in when done